include/oneapi/dpl/internal/dynamic_selection_impl/sycl_scheduler.h

```diff
@@ -0,0 +1,136 @@
+    using universe_container_t = std::vector<execution_resource_t>;
+
+    class async_wait_t {
+    public:
+      virtual void wait_for_all() = 0;
+      virtual native_sync_t get_native() const = 0;
+      virtual ~async_wait_t() {}
+    };
+    using waiter_container_t = Queue<async_wait_t *>;
+
+    template<typename PropertyHandle>
+    class async_wait_impl_t : public async_wait_t {
+      PropertyHandle p_;
+      native_sync_t w_;
+      std::chrono::high_resolution_clock::time_point t0_;
+      std::shared_ptr<std::atomic<bool>> wait_reported_;
+    public:
+
+      async_wait_impl_t(PropertyHandle p, sycl::event e) : p_(p), w_(e), t0_(std::chrono::high_resolution_clock::now()),
+                                                           wait_reported_{std::make_shared<std::atomic<bool>>(false)} { };
+
+      native_sync_t get_native() const override {
+        return w_;
+      }
+
+      void wait_for_all() override {
+        w_.wait();
+      }
+    };
+
+    std::mutex global_rank_mutex_;
+    universe_container_t global_rank_;
+    waiter_container_t waiters_;
+
+    sycl_scheduler() { }
+
+    sycl_scheduler(const sycl_scheduler& v) : global_rank_(v.global_rank_), waiters_(v.waiters_) { }
+
+    template<typename NativeUniverseVector, typename ...Args>
+    sycl_scheduler(const NativeUniverseVector& v, Args&&... args) {
+      for (auto e : v) {
+        global_rank_.push_back(e);
+      }
+    }
+
+    template<typename SelectionHandle, typename Function, typename ...Args>
+    auto submit(SelectionHandle h, Function&& f, Args&&... args) {
+      using PropertyHandle = typename SelectionHandle::property_handle_t;
+      auto w = new async_wait_impl_t<PropertyHandle>(h.get_property_handle(), f(h.get_native(), std::forward<Args>(args)...));
+      waiters_.push(w);
+      return *w;
+    }
+
+    void wait_for_all() {
+      async_wait_t *w;
+      waiters_.pop(w);
+      w->wait_for_all();
+      delete w;
+    }
+
+    //
+    // Support for property queries
+    //
+
+    auto query(oneapi::dpl::experimental::property::universe_t) noexcept {
+      std::unique_lock<std::mutex> l(global_rank_mutex_);
+      if (global_rank_.empty()) {
+        auto devices = sycl::device::get_devices();
+        for(auto x : devices){
+          global_rank_.push_back(sycl::queue{x});
+        }
+      }
+      return global_rank_;
+    }
+
+    auto query(oneapi::dpl::experimental::property::universe_size_t) noexcept {
+      if (global_rank_.empty()) {
+        query(oneapi::dpl::experimental::property::universe);
+      }
+      {
+        std::unique_lock<std::mutex> l(global_rank_mutex_);
+        return global_rank_.size();
+      }
+    }
+
+    auto report(oneapi::dpl::experimental::property::universe_t, const universe_container_t &gr) noexcept {
+      std::unique_lock<std::mutex> l(global_rank_mutex_);
+      global_rank_ = gr;
+    }
```

Comment:
It seems to me that replacing the universe during the scheduler lifetime via the property mechanism at the very least abuses that mechanism, and I am not sure if it is a good idea overall. Should it be removed before we figure out all nuances, or is it important to keep for the currently considered usage scenarios?

Reply:
Let's revisit after beta.